### PR TITLE
Dependency injection and tests in the new framework

### PR DESCRIPTION
**Pull request by themarolt**
# Changes proposed ✍️

### <samp>🤖 Generated by Copilot at f9840b2</samp>

This pull request improves the integration with Discourse forums, adds a new feature to install dependencies from the command line, updates some service dependencies and scripts, and simplifies the Dockerfile for the script executor service.
​

### <samp>🤖 Generated by Copilot at f9840b2</samp>

> _We're building the script executor, me hearties, yo ho_
> _We're cloning and installing the crowd.dev repo_
> _We're adding new features and fixing bugs, me lads and lasses, yo ho_
> _We're running the trigger-webhook script on the count of three_

### <samp>🤖 Generated by Copilot at f9840b2</samp>

*  Add script to trigger webhook for testing purposes to `backend/package.json` ([link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-495707834ca4b862f9acdfbac70d279023d2c059da13db59594e61ed3354fed5R38))
*  Add conditions to check if user is defined before proceeding with Discourse integration operations in `discourseIntegrationService.ts` ([link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-895e6815af572b0f428613cb171893e5bdcfd5ae2d2a8f988efd10f4703dc5a4R224-R229), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-895e6815af572b0f428613cb171893e5bdcfd5ae2d2a8f988efd10f4703dc5a4R421-R426), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-895e6815af572b0f428613cb171893e5bdcfd5ae2d2a8f988efd10f4703dc5a4R545-R550))
*  Modify return type, encode url parameter and handle 404 error response in `getUser.ts` ([link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-48b12a31632d2269f418fbe87f73f9f4cc4a42852edcf59f87fdb183b8827cc5L13-R13), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-48b12a31632d2269f418fbe87f73f9f4cc4a42852edcf59f87fdb183b8827cc5L21-R21), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-48b12a31632d2269f418fbe87f73f9f4cc4a42852edcf59f87fdb183b8827cc5R36-R39))
*  Add function to install backend and services libraries dependencies and handle install-backend-libs argument in `scripts/cli` ([link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-5e2da9ab3fa1cd8081cbd3848fb84fbe0fd7ce37b42a726520e9d10398a40ef2R142-R149), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-5e2da9ab3fa1cd8081cbd3848fb84fbe0fd7ce37b42a726520e9d10398a40ef2R605-R608))
*  Replace content of `Dockerfile.script_executor` with simplified Dockerfile using node:16-alpine base image and installing necessary packages and libraries ([link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-848393c1ddfe6af62978e066001dbc4fb8323d3d2e99073f82df91a6911349eaL3-R8))
*  Add command-line-args and command-line-usage dependencies and name properties to ../../libs dependencies in `integration_stream_worker/package.json` and `package-lock.json` ([link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bR20-R21), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bL38-R45), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bR60), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bR79), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bR106), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bR127), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bR147), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bL145-R155), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bR169), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bR182-R201), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-3831ba00f8d2658b222de90c364557d3fe41a692351de3b8fca48e7831d57e5fR35-R36), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-3831ba00f8d2658b222de90c364557d3fe41a692351de3b8fca48e7831d57e5fL42-R46))
*  Remove dev properties, add new dependencies and avoid duplication of dependencies in `integration_stream_worker/package-lock.json` ([link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bL670), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bR735-R742), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bL764), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bL776-R830), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bL845), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bL856-R950), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bR1411-R1421), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bL1461), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bR1774-R1783), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bR2319-R2326), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bR2388-R2423), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bR2581-R2588), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bR2632-R2639), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bR2727-R2742), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bR2770), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bR2858), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bL2992), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bR3207-R3211), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bL3065), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bL3075), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bR3274-R3281), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bL3124), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bL3132-R3366), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bR3706-R3713), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bL3575-R3815), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bR3975-R3984), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bR4356-R4360), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bR4401-R4426), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bR4530-R4534), [link](https://github.com/CrowdDotDev/crowd.dev/pull/1541/files?diff=unified&w=0#diff-8cdeb3d9ebdca74323fe3cf391204b44f2d44825eef460364ce5f786e1e3921bR4569-R4573))

## Checklist ✅
- [x] Label appropriately with `Feature`, `Improvement`, or `Bug`.
- [ ] ~Add screehshots to the PR description for relevant FE changes~
- [x] New backend functionality has been unit-tested.
- [ ] ~API documentation has been updated (if necessary) (see [docs on API documentation](https://docs.crowd.dev/docs/updating-api-documentation)).~
- [x] [Quality standards](https://github.com/CrowdDotDev/crowd-github-test-public/blob/main/CONTRIBUTING.md#quality-standards) are met.

### CODE REVIEW

**Comment by sausage-todd** [comment]
I'll put the comment here to start a discussion.

Thanks for all this effort Uroš, I have a couple issues though:
1. I'm still not sure how to review the PR, there are so many changes
2. It would've been much easier to review it if the changes were split into several reasonable commits, each having a specific purpose with an explanation (i.e. `Remove build step from libs`, `Add basic IOC setup`, `Migrate all classes to the IOC framework`)
3. It seems like there are still two separate things: no build step for libraries, and ioc. Can we split the PR even further?
4. Skeptical a bit about ioc. I think it adds too much complexity, and we're getting too close to java. I'm afraid instead of creating features we'll be creating interfaces and binding implementation to them.

**Comment by themarolt** [comment]
> I'll put the comment here to start a discussion.
> 
> Thanks for all this effort Uroš, I have a couple issues though:
> 
> 1. I'm still not sure how to review the PR, there are so many changes
> 2. It would've been much easier to review it if the changes were split into several reasonable commits, each having a specific purpose with an explanation (i.e. `Remove build step from libs`, `Add basic IOC setup`, `Migrate all classes to the IOC framework`)
> 3. It seems like there are still two separate things: no build step for libraries, and ioc. Can we split the PR even further?
> 4. Skeptical a bit about ioc. I think it adds too much complexity, and we're getting too close to java. I'm afraid instead of creating features we'll be creating interfaces and binding implementation to them.

I'll split the PR into two - DI and no lib building.

As for your 4th point:
- I think that DI will help with testing and scripting:
  - we will be able to inject custom script classes instead of predefined service classes like sqs emitters for example
  - as for the complexity - library code is getting more complex and with the plan to use library and service code in other applications this will help with that - user will be able to more easily initialize the library code and have it ready wherever we need it
  - I would argue on the contrary - it decreases complexity for the developer:
    - he doesn't need to know the initialization steps for the library components
    - he doesn't need to refactor multiple files if he want's to get a certain component into his constructor
    - inversify offers explicit bindings/mappings unlike some DI systems - I also dislike Java/Spring style DI but this one I like because it's not just wiring things seemingly like magic - it's explicit 
  - argument that we are too close to Java doesn't really hold IMO - Java has been an enterprise solution used for decated before NodeJS came knocking - it's one of the most stable languages/ecosystems there are - if you are implying Java with Spring as DI then yes I can relate - I dislike that a lot as well :)

**Comment by themarolt** [comment]
@sausage-todd - PR for [no lib building](https://github.com/CrowdDotDev/crowd.dev/pull/1624)

File: services/tests/jest.config.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
+/* eslint-disable @typescript-eslint/no-var-requires */
+
+const tsconfig = require('./tsconfig.json');
+
+const fromPairs = (pairs) =>
+  pairs.reduce((res, [key, value]) => ({ ...res, [key]: value }), {});
+
+/**
+ * tsconfig の paths の設定から moduleNameMapper を生成する
```
Comment (epipav) [inline services/tests/jest.config.js:9]: NIT: let's remove the non-latin comments

**Review by epipav — APPROVED** [review]
lgtm :+1:  added 3 nitpicks

**Pull request closed without merging**